Users, Sign-In & Leader request flow

ranyitz opened the pull request:
This PR is work in progress and should solve the issue described in #70 

At the moment it focuses on the logic and the structure, but not on the design. The design can be added in a later PR. At the moment there is no way to get to the sign-up flow from the UI. (You can sign up using the `/sign-up` route) but there is not harm that can be done.

## Summary

### Sign up flow
1. By navigating to `/sign-up?protest=hA0gksQ18Ib6FiLpy8df` we're getting to the sign-up page.
2. In the sign up page there is a login with facebook button, when pressed we're being redirected to the `/leader-request?protest=hA0gksQ18Ib6FiLpy8df`

> Note that we're passing the query parameter of the protest id

3. In the leader request page, we need to add the phone number, after pressing on the `submit page` we'll move to `/leader-request-submitted`

### Log out
Users can log out from the top left corner

### Profile page
You can press on your name on the top left corner in order to get to the profile page (The profile page is just a placeholder)

## What's left to be done?
1. Design and implementation of all of the views
* [ ] sign up page
* [ ] leader request page
* [ ] leader request submitted page
* [ ] profile page
* [ ] header logout button and profile link
2. Validation for correct phone number before sending.
3. complete the flow by approving the requests from the admin page @daganomri 

In the `protests` collection, we'll need to add the leader role with the right user ids, I added all of the functions that are needed in `SignUp/utils.js`

```js
{
 "roles":
    {
    "leader": ['fieu2h37heiufh784r']
    }
}
```

## Structure

### `leader_requests` collection
We're going to add a request every time a new request is submitted.
* `status: 'pending'/'done'`
* protestId
* User data

### `users` collection
We need this collection in order to save data which we don't have on the object that we get from firebase/facebook, later on we'll add role/admin property on this user object.

ranyitz committed:
wip

ranyitz committed:
new flow

ranyitz committed:
small improvements

ranyitz committed:
more fixes

daganomri commented:
Will the sign-up page work without the protestId? In the case of a new user that wants to create a protest, we'll want him to sign up before creating the protest, right?

webernir commented:
Sign-up and Leadership are 2 different concepts and should be separated. Meaning, sign-up should know nothing about `protestId`.

When a user want to claim a leadership role:
- if already logged-in - go to `/leader-request?protest={ID}`
- If not logged-in - go to `/sign-up` page. We can also add `?returnUrl` so user can be redirected back after sign-up.

We can also handle this on the route-level by adding `onRouteEnter` guard to `/leader-request` 

This will also address @daganomri comment

What do you think @ranyitz ?

ranyitz commented:
@daganomri @webernir both of you are right. I actually thought the same, after consulting with @guytepper he said that at the moment this will be the main use-case.

In order to support sign up without redirect to claim leadership on a protest we should:
1. After sign up is done, only move to `/leader-request` route in case there is a query parameter of `?protest={PROTEST_ID}`.
2. Create a finish registration view.

Claiming leadership on a protest for registered users already theoretically works and only assume that the user needs to be registered before claiming the leadership. So the when pressing on "claim leadership" we need to understand whether the user is logged in or not and send him/her to either sign-up or to claim leadership.

ranyitz commented:
@webernir I also agree that we can have a `returnUrl` with the full url of `/leader-request?protest=hA0gksQ18Ib6FiLpy8df
` so the sign up will just redirect to this url when the registration is done. I'll change the implementation.

ranyitz committed:
return-url

ranyitz commented:
I changed the implementation to work with `/sign-up?returnUrl=/leader-request?protest=hA0gksQ18Ib6FiLpy8df`. This means that the protest page needs to go to this URL in order to signUp + return to the leader-request page.

I also rebased the changes from master so this PR should be green.

uriklar commented on src/App.js:
```diff
@@ -13,6 +24,17 @@ import { setLocalStorage, getLocalStorage } from './localStorage';
 
 const GeoFirestore = geofirestore.initializeApp(firestore);
 
+async function signOut() { 
+  firebase.auth().signOut().then(function() {
+  }, function(error) {
+    console.error(error);
+  });
+}
+
+async function getFullUserData(uid) {
+  return (await firestore.collection('users').doc(uid).get()).data();
+}
+
```
Can we move these to the api/index.js file?

uriklar commented on src/App.js:
```diff
@@ -214,6 +256,17 @@ function App() {
+            <Route exact path="/leader-request">
+              <LeaderRequest user={state.user}/>
+            </Route>
+            <Route exact path="/leader-request-submitted">
+              <LeaderRequestSubmitted user={state.user}/>
+            </Route>
```
Maybe this could be a nested route inside LeaderRequest? leader-request/submitted. And this route can live inside that file

ranyitz replied:
Moved to be a nested route 👍

uriklar commented on src/views/LeaderRequest.js:
```diff
@@ -0,0 +1,63 @@
+  const history = useHistory();
+
+  async function submitLeaderRequest(phoneNumber) {
+    const { protest: protestId } = queryString.parse(window.location.search);
```
I guess it's not critical but you can make this a react-router param and use it like I did with the protestId in the protest page

ranyitz replied:
We use `query` and not param here. You suggest to switch to a param?

uriklar commented on src/views/Profile.js:
```diff
@@ -0,0 +1,50 @@
+      return;
+    }
+
+    var protestsRef = firestore.collection('protests');
+    var query = protestsRef.where('roles.leader', 'array-contains', props.user.uid);
+  
+    query.get()
+    .then(function(querySnapshot) {
+      const protests = [];
+
+      querySnapshot.forEach(function(doc) {
+        protests.push({ id: doc.id, ...doc.data() });
+      });
```
I'd also move this into api/index.js

uriklar commented on src/views/SignUp/index.js:
```diff
@@ -0,0 +1,68 @@
+import queryString from 'query-string';
+
+function SignUpBeforeRedirect() {
+  return (
+    <>
+    <Button onClick={() => handleSignIn()}>Sign Up Using Facebook</Button>
+    </>
```
Can we put this in the header? instead of it being a separate page

ranyitz replied:
Maybe, but this would be a different product, I think that @guytepper doesn't want to encaurage anybody to sign-up, only people that want to claim ownership (be the leader) of a protest.

uriklar commented on src/views/SignUp/utils.js:
```diff
@@ -0,0 +1,102 @@
+import firebase, { firestore } from '../../firebase';
```
I think all these functions should be in the api file. That way we have one place the concentrates all calls to firestore

ranyitz replied:
Agree, i'll move them

ranyitz committed:
fix merge bug

ranyitz committed:
move logic into api

ranyitz committed:
add submitted to be a nested route under leader request

ranyitz committed:
add phone number to user

guytepper commented:
Although the PR author won't agree, I think this PR is a work of art that was done in less than 8 hours. 
 
Thanks everyone who took part and helped review it.

ranyitz commented:
The url that needs to be used from the protest page - `/sign-up?returnUrl=/leader-request?protest=hA0gksQ18Ib6FiLpy8df`
